File: crates/core/src/string/transform.rs
// transform.rs
// String transformation utilities

pub fn reverse(input: &str) -> String {
    input.chars().rev().collect()
}

pub fn replace(input: &str, search: &str, replacement: &str) -> String {
    input.replace(search, replacement)
}

pub fn slugify(input: &str, separator: char) -> String {
    let mut out = String::with_capacity(input.len());
    let mut prev_sep = false;

    for ch in input.trim().chars().flat_map(|c| c.to_lowercase()) {
        if ch.is_alphanumeric() {
            out.push(ch);
            prev_sep = false;
        } else if ch.is_whitespace() && !prev_sep {
            out.push(separator);
            prev_sep = true;
        }
    }

    // Remove trailing separator if present
    if out.ends_with(separator) {
        out.pop();
    }

    out
}

File: webapp/src/lib/utils/string/transform.ts
import type { ToolResult } from "../../../types";
import { wasmWrapper } from "../../wasm";

export const stringTransformOperations = {
  reverse: (input: string): ToolResult => {
    return wasmWrapper.string_reverse(input);
  },
  replace: (input: string, search: string, replacement: string): ToolResult => {
    return wasmWrapper.string_replace(input, search, replacement);
  },
  slugify: (input: string, separator = "-"): ToolResult => {
    return wasmWrapper.string_slugify(input, separator);
  },
};
